### PR TITLE
[Editor] Enable multi-editing of Cinemachine components/extensions, where applicable

### PR DESCRIPTION
**Pull request by AscendingMan**
-----
### Purpose of this PR

Added multi-selection support to Cinemachine editors where it made sense, including:

- Virtual Camera
- Free Look Camera
- Blend List, Clear Shot Cameras (support is limited due to these pointing to child objects of the GO to which the camera component is attached, thus there is no multi-editing for that)
- All of the Extensions
------
### Testing status

Tested manually to work when handling multiple objects by myself and our internal team QA.

### Screenshots
CinemachineMultiEdti

### CODE REVIEW

**Comment by AscendingMan** [comment]
Cc @glabute @aras-p For exposure as I can't add reviewers for some reason >_<

File: Editor/Editors/CinemachineVirtualCameraBaseEditor.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 using System;
 using System.Collections.Generic;
 using Cinemachine.Utility;
+using UnityEngine.UIElements;
```
Comment (aras-p) [inline Editor/Editors/CinemachineVirtualCameraBaseEditor.cs:6]: Minor: why using of UIElements namespace is needed here?

File: Editor/Editors/CinemachineVirtualCameraBaseEditor.cs [hunk 2]
```diff
@@ -111,21 +118,37 @@ protected void DrawExtensionsWidgetInInspector()
...
+                rect = EditorGUI.PrefixLabel(rect, Styles.addExtensionLabel);
 
                 int selection = EditorGUI.Popup(rect, 0, sExtensionNames);
                 if (selection > 0)
                 {
                     Type extType = sExtensionTypes[selection];
-                    if (Target.GetComponent(extType) == null)
-                        Undo.AddComponent(Target.gameObject, extType);
+
+                    if (Selection.objects.Length == 1)
```
Comment (aras-p) [inline Editor/Editors/CinemachineVirtualCameraBaseEditor.cs:128]: Why this needs separate code paths for 1 vs non-1 length? From a quick look it feels like the "loop over targets" would just work fine for 1 length as well?

File: Editor/Editors/CinemachineVirtualCameraEditor.cs [hunk 3]
```diff
@@ -40,6 +42,14 @@ protected override void OnEnable()
             base.OnEnable();
             IsPrefab = Target.gameObject.scene.name == null; // causes a small GC alloc
             UpdateStaticData();
+            UpdateStageDataTypeMatchesForMultiSelection();
+            Undo.undoRedoPerformed += ResetTargetOnUndo;
```
Comment (aras-p) [inline Editor/Editors/CinemachineVirtualCameraEditor.cs:46]: Does this ever de-register undoRedoPerformed handler?

File: Editor/Editors/CinemachineVirtualCameraEditor.cs [hunk 4]
```diff
@@ -390,9 +443,21 @@ void UpdateComponentEditors()
...
+                        if (cinemachineVirtualCamera == null)
+                            continue;
+
+                        var behaviour = cinemachineVirtualCamera.GetCinemachineComponent(components[i].Stage) as MonoBehaviour;
+                        if (behaviour != null)
+                            behaviours.Add(behaviour);
+                    }
+
+                    var behaviourArray = behaviours.ToArray();
+                    if (behaviourArray.Length > 0 && behaviourArray[0] != null && (hasSameStageDataTypes[(int)components[i].Stage] || targets.Length == 1))
```
Comment (aras-p) [inline Editor/Editors/CinemachineVirtualCameraEditor.cs:459]: Why the `behaviourArray[0] != null` check is needed? I don't quite see how would a null entry get in there

File: Editor/Editors/CinemachineVirtualCameraBaseEditor.cs (hunk 1 above)
Comment (AscendingMan) [reply Editor/Editors/CinemachineVirtualCameraBaseEditor.cs:6]: Rider accident.

File: Editor/Editors/CinemachineVirtualCameraBaseEditor.cs (hunk 2 above)
Comment (AscendingMan) [reply Editor/Editors/CinemachineVirtualCameraBaseEditor.cs:128]: I missed that when refactoring, don't think there's any other reason to it.

File: Editor/Editors/CinemachineVirtualCameraEditor.cs (hunk 4 above)
Comment (AscendingMan) [reply Editor/Editors/CinemachineVirtualCameraEditor.cs:459]: Yep, leftovers from when I used to cast it to a different type. Will fix.

File: Editor/Editors/CinemachineVirtualCameraBaseEditor.cs [hunk 5]
```diff
@@ -13,6 +14,12 @@ namespace Cinemachine.Editor
     public class CinemachineVirtualCameraBaseEditor<T>
         : BaseEditor<T> where T : CinemachineVirtualCameraBase
     {
+        public static class Styles
+        {
+            public static GUIContent addExtensionLabel = new GUIContent("Add Extension");
+            public static GUIContent virtualCameraChildrenInfoMsg = new GUIContent("The Virtual Camera Children field displays the child objects of a given parent object, thus this field is not available when multiple objects are selected.");
```
Comment (AscendingMan) [inline Editor/Editors/CinemachineVirtualCameraBaseEditor.cs:20]: This text will need to be fixed not to display "Virtual Camera" for cases when it's on a different type of camera.

File: Editor/Editors/CinemachineVirtualCameraEditor.cs [hunk 6]
```diff
@@ -153,16 +163,28 @@ protected void DrawPipelineInInspector()
...
                 r = rect; r.width -= labelWidth; r.x += labelWidth;
                 GUI.enabled = !StageIsLocked(stage);
-                int newSelection = EditorGUI.Popup(r, m_stageState[index], sStageData[index].PopupOptions);
+
+                if (targets.Length > 1)
+                    EditorGUI.showMixedValue = !hasSameStageDataTypes[index];
+
                 GUI.enabled = true;
+                EditorGUI.BeginChangeCheck();
+                int newSelection = EditorGUI.Popup(r, m_stageState[index], sStageData[index].PopupOptions);
```
Comment (glabute) [inline Editor/Editors/CinemachineVirtualCameraEditor.cs:172]: GUI.enabled = !StageIsLocked(stage) above is supposed to apply to the popup - you have broken that by setting GUI.enabled = true.

FreeLook, for example, does not allow the user to change the popup value for Body

image

File: Editor/Editors/CinemachineVirtualCameraEditor.cs [hunk 7]
```diff
@@ -382,25 +378,57 @@ bool useSimple
             }
         }
 
+        List<CinemachineComponentBase> GetSortedPipelineComponentsForCamera(UnityEngine.Object targetObject)
+        {
+            var camera = targetObject as CinemachineVirtualCamera;
+            if (camera == null)
+                return null;
+
+            var cameraComponents = new List<CinemachineComponentBase>();
```
Comment (glabute) [inline Editor/Editors/CinemachineVirtualCameraEditor.cs:387]: This code is creating a lot of GC allocs, which are very bad for editor performance.

Comment (glabute) [reply Editor/Editors/CinemachineVirtualCameraEditor.cs:387]: You can replace the whole thing with this, which is much more efficient:

```
        void GetPipelineTypes(CinemachineVirtualCamera vcam, ref Type[] types)
        {
            for (int i = 0; i < types.Length; ++i)
                types[i] = null;
            if (vcam != null)
            {
                var components = vcam.GetComponentPipeline();
                for (int j = 0; j < components.Length; ++j)
                    types[(int)components[j].Stage] = components[j].GetType();
            }
        }

        // scratch buffers for pipeline types
        Type[] s_PipelineTypeCache0 = new Type[Enum.GetValues(typeof(CinemachineCore.Stage)).Length];
        Type[] s_PipelineTypeCacheN = new Type[Enum.GetValues(typeof(CinemachineCore.Stage)).Length];

        void UpdateStageDataTypeMatchesForMultiSelection()
        {
            for (int i = 0; i < m_hasSameStageDataTypes.Length; ++i)
                m_hasSameStageDataTypes[i] = true;

            if (targets.Length > 1)
            {
                GetPipelineTypes(serializedObject.targetObjects[0] as CinemachineVirtualCamera, ref s_PipelineTypeCache0);
                for (int i = 1; i < targets.Length; ++i)
                {
                    GetPipelineTypes(serializedObject.targetObjects[i] as CinemachineVirtualCamera, ref s_PipelineTypeCacheN);
                    for (int j = 0; j < s_PipelineTypeCache0.Length; ++j)
                        if (s_PipelineTypeCache0[j] != s_PipelineTypeCacheN[j])
                            m_hasSameStageDataTypes[j] = false;
                }
            }
        }
```

**Comment by glabute** [comment]
Interesting.  I am also on VS 2019, and I get this warning.  Maybe you should check if your VS setup is correct?

**Comment by aras-p** [comment]
Would this need a changelog entry btw?

**Comment by glabute** [comment]
yeah, I'll take care of it

**Pull request merged**